### PR TITLE
Feature/inject: Select application to run, inject wrapper (linux only)

### PR DESCRIPTION
**Pull request by mrzzzrm**

### CODE REVIEW

File: source/client/CMakeLists.txt [hunk 1]
```diff
@@ -27,7 +27,8 @@ set(sources
     CloudGLClient.h
     CreateSessionDialog.cpp
     CreateSessionDialog.h
-)
+        TraceProcess.cpp
+        TraceProcess.h)
```
Comment (apopiak) [inline source/client/CMakeLists.txt]: inconsistent indentation

File: source/client/CloudGLClient.cpp [hunk 2]
```diff
@@ -13,7 +15,22 @@ CloudGLClient::CloudGLClient()
 }
 
 void CloudGLClient::createSession() {
+    if (m_traceProcess) {
+        std::cerr << "Can't create session, trace process already running" << std::endl;
+        return;
+    }
+
```
Comment (apopiak) [inline source/client/CloudGLClient.cpp]: do we use `std::cerr` and `std::cout` or `qDebug()` etc.?

File: source/client/CreateSessionDialog.cpp [hunk 3]
```diff
@@ -2,13 +2,39 @@
...
+
+}
 
 CreateSessionDialog::CreateSessionDialog(QWidget * parent):
     QDialog(parent)
 {
     m_ui.reset(new Ui::CreateSessionDialog());
     m_ui->setupUi(this);
+
+    // Read first than set, otherwise setting the app path will override the saved
```
Comment (apopiak) [inline source/client/CreateSessionDialog.cpp]: minor type: "Read first then set

File: source/client/CreateSessionDialog.cpp [hunk 4]
```diff
@@ -21,16 +47,37 @@ const std::string & CreateSessionDialog::workingDirectory() const
...
     m_applicationPath = path;
     m_ui->applicationPathLineEdit->setText(QString::fromStdString(path));
+    SETTINGS.setValue(APPLICATION_PATH_KEY, QString::fromStdString(path));
 }
 
 void CreateSessionDialog::setWorkingDirectory(const std::string & directory)
 {
     m_workingDirectory = directory;
-    m_ui->workingDirectoryLineEdit->setText(QString::fromStdString(directory));
+    m_ui->workingDirectoryLineEdit->setText(QString::fromStdString(directory));;
```
Comment (apopiak) [inline source/client/CreateSessionDialog.cpp]: minor: double semicolon

File: source/client/CreateSessionDialog.cpp [hunk 5]
```diff
@@ -21,16 +47,37 @@ const std::string & CreateSessionDialog::workingDirectory() const
...
+    SETTINGS.setValue(WORKING_DIRECTORY_KEY, QString::fromStdString(directory));
+}
+
+void CreateSessionDialog::setArguments(const std::string & arguments)
+{
+    m_arguments = arguments;
+    m_ui->argumentsLineEdit->setText(QString::fromStdString(arguments));;
```
Comment (apopiak) [inline source/client/CreateSessionDialog.cpp]: minor double semicolon

File: source/client/CreateSessionDialog.cpp [hunk 6]
```diff
@@ -41,7 +88,22 @@ void CreateSessionDialog::on_applicationPathButton_clicked(bool /*checked*/)
...
 
+    if (fileName.isNull()) {
+        return;
+    }
+
+    if (!fileName.isEmpty()) {
+        if (!isFileOk(fileName.toStdString())) {
+            return;
+        }
+    }
```
Comment (apopiak) [inline source/client/CreateSessionDialog.cpp]: no errors here?

Comment (apopiak) [inline source/client/CreateSessionDialog.cpp]: ok, found the messagebox warnings, still: no error on `filename.isNull()`?

File: source/client/TraceProcess.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            this, SLOT(traceStarted()));
+}
+
+void TraceProcess::start()
+{
+    auto wrapperFilename = "glxtrace.so";
+
+    auto wrapperPath = findWrapper(wrapperFilename, true);
+    if (wrapperPath.isEmpty()) {
+        qDebug() << "error: failed to find " << wrapperFilename << " wrapper (rerun with -v option for more details)\n";
```
Comment (apopiak) [inline source/client/TraceProcess.cpp]: different error level? `qWarning`?

File: source/client/TraceProcess.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    QStringList arguments;
+    if (!m_args.empty()) {
+        arguments << QLatin1String(m_args.c_str());
+    }
+
+    m_process->setWorkingDirectory(QString::fromStdString(m_workingDir));
+    m_process->start(QString::fromStdString(m_applicationPath), arguments);
+
+    std::cout << m_applicationPath << " " << m_args << std::endl;
+    qDebug() << QString::fromStdString(m_applicationPath) << " " << arguments;
```
Comment (apopiak) [inline source/client/TraceProcess.cpp]: `std::cout` and `qDebug()` in 2 lines? :-D

File: source/client/TraceProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+void TraceProcess::processFinished()
+{
+#if 1
```
Comment (apopiak) [inline source/client/TraceProcess.cpp]: purpose?

File: source/client/TraceProcess.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+void TraceProcess::processError(QProcess::ProcessError err)
+{
+#if 1
+    qDebug() << "trace error = " << m_applicationPath.c_str() << "(" << err << ")";
+    qDebug() << "\terr = " << m_process->readAllStandardError();
+    qDebug() << "\tout = " << m_process->readAllStandardOutput();
+#endif
+    //emit error(m_process->readAllStandardError());
```
Comment (apopiak) [inline source/client/TraceProcess.cpp]: remove outcommented code?

File: source/client/main.cpp [hunk 11]
```diff
@@ -16,18 +16,18 @@ using namespace common;
...
+//    SharedMemoryDevice mem(static_cast<size_t>(512));
+//    std::vector<char> data(512, 0);
+//    if (mem.open(QIODevice::ReadWrite)) {
+//        while (data[0] == 0) {
+//            mem.read(data.data(), 512);
+//            std::this_thread::sleep_for(std::chrono::milliseconds(50));
+//        }
+//    }
+//
+//    std::cout << "Client says: " << std::endl << data.data() << std::endl;
```
Comment (apopiak) [inline source/client/main.cpp]: :-D

**Comment by apopiak** [comment]
mostly looks good to me, let's talk about error messages ;-)

File: source/client/CloudGLClient.cpp (hunk 2 above)
Comment (mrzzzrm) [inline source/client/CloudGLClient.cpp]: In general I would say: std::cout for logging, std::cerr for errors. And when there are a lot of QStrings around fall back to qDebug()/qWarning()/... just for the convenience of not having to type ```.toStdString()``` all the time.

File: source/client/CreateSessionDialog.cpp (hunk 6 above)
Comment (mrzzzrm) [inline source/client/CreateSessionDialog.cpp]: ```filename.isNull()``` should be the case when the user cancels the ```QFileDialog::getOpenFileName```-dialog.

File: source/client/main.cpp (hunk 11 above)
Comment (mrzzzrm) [inline source/client/main.cpp]: This is your testing code. Do you want it removed?

**Comment by mrzzzrm** [comment]
All fine?

**Pull request merged**